app/clients/sms/twilio.py

```diff
@@ -110,12 +151,66 @@ def send_sms(self, to, content, reference, **kwargs) -> str:
 
                 self.logger.info(f"Twilio message created using messaging_service_sid")
 
-            self.logger.info(f"Twilio send SMS request for {reference} succeeded: {message.sid}")
+            self.logger.info(
+                "Twilio send SMS request for %s succeeded: %s", reference, message.sid
+            )
 
             return message.sid
         except Exception as e:
-            self.logger.error(f"Twilio send SMS request for {reference} failed")
+            self.logger.error("Twilio send SMS request for %s failed", reference)
             raise e
         finally:
             elapsed_time = monotonic() - start_time
-            self.logger.info(f"Twilio send SMS request for {reference} finished in {elapsed_time}")
+            self.logger.info(
+                "Twilio send SMS request for %s  finished in %s",
+                reference,
+                elapsed_time,
+            )
+
+    def translate_delivery_status(self, twilio_delivery_status_message) -> dict:
+        """
+        Parses the base64 encoded delivery status message from Twilio and returns a dictionary.
+        The dictionary contains the following keys:
+        - record_status: the convereted twilio to notification platform status
+        - reference: the message id of the twilio message
+        - payload: the original payload from twilio
+        """
+        if not twilio_delivery_status_message:
+            raise Exception("Twilio delivery status message is empty")
+
+        decoded_msg = base64.b64decode(twilio_delivery_status_message).decode()
+
+        parsed_dict = parse_qs(decoded_msg)
+
+        if "MessageStatus" not in parsed_dict:
+            raise Exception("Twilio delivery status message is missing MessageStatus")
+
+        twilio_delivery_status = parsed_dict["MessageStatus"][0]
+
+        if twilio_delivery_status not in twilio_notify_status_map:
+            raise Exception("Invalid Twilio delivery status:", twilio_delivery_status)
+
+        if "ErrorCode" in parsed_dict and (
+            twilio_delivery_status == "failed"
+            or twilio_delivery_status == "undelivered"
+        ):
+            error_code = parsed_dict["ErrorCode"][0]
+
+            if error_code in twilio_error_code_map:
+                notify_delivery_status = twilio_error_code_map[error_code]
+            else:
+                notify_delivery_status = twilio_notify_status_map[
+                    twilio_delivery_status
+                ]
+        else:
```

Comment:
This does not cover a cancel status. Also, if any of the other ones throw an error code but are not covered it'd come here. Under what circumstances will Twilio have an error code but be anything but failed, undelivered, or canceled?

Reply:
error codes only occur for failed and undelivered.  From Twilio Docs:
```
When a message's status is failed or undelivered, the ErrorCode and ErrorMessage properties will contain one of the following.
```